### DISCUSSION TITLE
Video support in Runtime

### DISCUSSION THREAD

**Discussion by twistedmedia** [airsdk/Adobe-Runtime-Support#847] [General]
Starting a thread here; our main use for Air is playback of various media (video and interactive/animated content from Animate) on tv and film sets. If you saw Godzilla vs Kong, I got a credit for the software I created that runs the computers and display on set. Anyhow...we REALLY would love to be able to move forward with the Harman version of the runtime but needs video support. Is this going to happen?

**Comment by ajwfrost** [comment]
Hi @twistedmedia - with the 33.1.1.476 build, videos on Windows should be fairly stable now; we've just had some updates go in on the MacOS side to reduce memory leakage so these will be in our next release.

But, for quite a while I had been hoping to completely rework the multimedia support in AIR because it's an incredibly complex area of code, having had a huge number of feature updates and process changed added in over the years: it's very hard to maintain, as we've been finding (and as everyone will have noticed...)

We're starting to get to the point where we have capacity to do this, and are planning to set up new classes to handle multimedia cross-platform. It would be good to check that folk would be able to update their apps to take advantage of these; we can try to implement the older APIs (NetStream etc) using these new classes, if that would be useful, but I'm thinking a clean break would actually be helpful too. So, e.g. a new set of classes under an `air.media` package that would be more graph based (like gstreamer or DirectShow), with the ability for anyone to create/add functionality into this too (either via special ANEs, or we did some RTSP handling a while that used Flascc).

So e.g. we define an interface for 'media sources' (and have implementations of these for getting media from local files, from embedded data, from http, rtmp, rtsp, dash etc), then we have some intermedia transform classes (demux, decode etc - most of the creation of these would probably need to be inferred), and then we have the 'sink' objects being the device sound output, video outputs including rgb rendering if needed (the traditional Video class), plus StageVideo or VideoTexture, etc.

Maybe we can use this conversation thread to discuss the plans and finalise the APIs.. it would be good to do this in collaboration with the folk who would actually use this!

**Reply by itlancer** [reply]
I agree with new APIs set "from scratch" approach without legacy maintain (except critical fixes such as crashes/huge memory leaks, non-working functionalities).
AIR should have modern mechanisms in such areas for further improvements.

@ajwfrost, I think you should create new thread for that discussion when you will be ready.
My basic points vision about AIR video/multimedia area have mentioned here https://github.com/airsdk/Adobe-Runtime-Support/issues/180#issuecomment-566217177

**Comment by twistedmedia** [comment]
To detail what kinds of things we'd like to be able to do (again):

• Play video above or behind other elements on the stage. For example, if video is below other elements, you can overlay a news banner lower 3rd or sci-fi targeting graphic etc. If video is above, it can fill an area in the background graphic, like a video feed or animated part of a larger graphic.

• Play MP4/H264+ locally

• Apply color transform (and other filters) to the video object. WE USE THIS 100% of the time...use case is "color correcting" on set. Often times the colors displayed on a monitor need to be adjusted to be 'seen' properly by the camera. From very minor tweaks ("it looks too magenta") to larger moves (dimming colors down to 10% for instance), this is a deal breaker if the functionality goes away.

**Comment by davecates** [comment]
Those features and functionalities would be epic...

One addition from me would be to ensure the capability to be able to add multiple video objects onto the stage simultaneously. We have had a nightmare doing three-way video streaming configs previously.

RTSP, RTMP and HLS would be needed by us as well.

What about the capture side as well with microphone and camera - compression and broadcasting?
